### PR TITLE
212 add support for oneof

### PR DESCRIPTION
**Pull request by Logende**
not perfect yet (for example the Dropdown selection text is not centered vertically), but functional

### CODE REVIEW

File: frontend/config-assistant/src/helpers/schema/schemaPreprocessor.ts [hunk 1]
```diff
@@ -63,3 +79,15 @@ function hasRef(schema: JsonSchemaObjectType): boolean {
...
+function hasOneOfs(schema: JsonSchemaObjectType): boolean {
+  return schema.oneOf !== undefined && schema.oneOf.length > 0;
+}
+function hasAnyOfs(schema: JsonSchemaObjectType): boolean {
+  return schema.anyOf !== undefined && schema.anyOf.length > 0;
+}
+
+function optimizeSchema(schema: JsonSchemaObjectType) {
+  if (hasOneOfs(schema)) {
+    // TODO: if it is just oneOf of types, then replace it by a choice of types
```
Comment (PaulBredl) [inline frontend/config-assistant/src/helpers/schema/schemaPreprocessor.ts:91]: There is a lot we can possible do here, not just types. We could check if the oneOf schemas are even relevant for the GUI editor

File: frontend/config-assistant/src/components/gui-editor/ConfigTreeNodeResolver.ts [hunk 2]
```diff
@@ -45,11 +45,20 @@ export class ConfigTreeNodeResolver {
...
       children: [],
-      leaf: this.isLeaf(schema, depth),
+      leaf: this.isLeaf(schema, absolutePath, depth),
     };
   }
 
-  private isLeaf(schema: JsonSchema, depth: number): boolean {
+  private isLeaf(schema: JsonSchema, absolutePath: Path, depth: number): boolean {
+    const dependsOnUserSelection = schema.anyOf.length > 0 || schema.oneOf.length > 0;
+    if (dependsOnUserSelection) {
```
Comment (PaulBredl) [inline frontend/config-assistant/src/components/gui-editor/ConfigTreeNodeResolver.ts:54]: for now its fine but actually we would need two different components: One for oneOf (single selection), one for multiple selection (anyOf) -> https://primevue.org/multiselect/

**Comment by Logende** [comment]
I also think we could add some Schema selection inference as initial selection: before the user has selected an oneOf option, none is selected currently. Instead, we could automatically check which Schema successfully validates the current data and choose the first one as initial selection.

File: frontend/config-assistant/src/components/gui-editor/PropertyMetadata.vue [hunk 3]
```diff
@@ -64,6 +76,12 @@ function getTypeDescription(): string {
   if (props.nodeData.schema.enum) {
     return 'enum';
   }
+  if (props.nodeData.schema.oneOf.length > 0) {
```
Comment (PaulBredl) [inline frontend/config-assistant/src/components/gui-editor/PropertyMetadata.vue:79]: I don't like that too much. AnyOf or OneOf isn't really a type

**Comment by PaulBredl** [comment]
image
It would be nice if the dropdown could have a max width so this doesn't happen

**Comment by PaulBredl** [comment]
In the strenda schema there is a oneOf / anyOf, but it doesn't allow me to select an option:
image

File: frontend/config-assistant/src/components/gui-editor/PropertyMetadata.vue (hunk 3 above)
Comment (Logende) [reply frontend/config-assistant/src/components/gui-editor/PropertyMetadata.vue:79]: this is something to discuss. It is not a type, but it has an explanatory effect and helps the user understand the dropdown menu. We need at least some way to show what the dropdown menu is about (or the multiselect for anyOf)

**Comment by PaulBredl** [comment]
wir können das so mergen, allerdings müsstest du noch einen konflikt beheben

**Review by PaulBredl — APPROVED** [review]

**Review by PaulBredl — APPROVED** [review]

**Pull request merged**